### ISSUE TITLE
[dart:async] inconsistent behavior with `_SyncBroadcastStreamController` and `_SyncStreamController()`  

### ISSUE THREAD

**Issue by beeth0ven** [dart-lang/sdk#47648]
Hello dart team!

I may found a bug with `StreamController`, We can reproduce it with following code:

## 1. with `StreamController.broadcast(sync: true)`:

```dart
void main() {
    late StreamController<int> _controller;

    _controller = StreamController<int>.broadcast( // <- broadcast stream
      sync: true,
      onListen: () {
        _controller.add(0);
      },
    );

    _controller.stream.listen((value) {
      print('next: $value');
    }); 

    print('listened');
}
```

will prints:

```diff
next: 0
listened
```

## 2. with `StreamController(sync: true)`:

```diff
void main() {
    late StreamController<int> _controller;

-  _controller = StreamController<int>.broadcast( // <- broadcast stream
+  _controller = StreamController<int>( // <- non broadcast stream
      sync: true,
      onListen: () {
        _controller.add(0);
      },
    );

    _controller.stream.listen((value) {
      print('next: $value');
    }); 

    print('listened');
}
```

will prints:

```
listened
next: 0
```

### output difference

1. sync broadcast stream cotroller
```diff
next: 0
listened
```

2. sync non broadcast stream cotroller
```
listened
next: 0
```

They have inconsistent behavior here. Since it's synchronous stream, I expect behavior 1 is correct. Behavior 2 may cause bug in [Rx.combineLatest](https://github.com/ReactiveX/rxdart/blob/a8ff5d1080b7a6d5a407bbcb4748da1b5051a63b/lib/src/streams/combine_latest.dart#L298-L301) operator in `rxdart`, we expect stream applied this operator should send synchronous event immediately.

What do you think about this?

Thanks!

## Environment

dart version: Dart SDK version: 2.13.3 (stable) (Wed Jun 9 12:44:44 2021 +0200) on "macos_x64"

**Comment by lrhn** [comment]
That's is a bug. The broadcast stream should not have printed anything.

A stream subscription isn't active until the `listen` call has returned. If you try to add events before that, which includes *during* the `onListen` callback, it should be treated as if the new subscription isn't active yet (because it isn't, the caller hasn't gotten access to the `StreamSubscription` object and might not have added any callbacks yet).

A non-broadcast controller *buffers* events that are added to the controller before a listener has subscribed. Those events are delivered *asynchronously* after the `listen` call has completed (it can't do it during, there is no listener yet, so it has to schedule a microtask to do it later). While events are buffered, even a `sync` stream controller cannot deliver events, it has to put them at the back of the buffer and wait for the prior events to be delivered, because events must be delivered in the correct order..
(Buffering was an unnecessarily complicated design, but it's been like that since the start. If designed today, it would likely have throw a `StateError` if you added events before you have a listener.)

A broadcast stream does not buffer events. If it has no listener at the point where the event is added, the event is just dropped (aka. immediately delivered to zero listeners). Any currently active listeners on a broadcast stream will be notified of the event when the event is added, and then deliver them as necessary (if it's a normal broadcast stream, they'll deliver it asynchronously, if `sync` they'll deliver it "immediately". I that quote because if there is more than one listener, they will take turns, so it's not atomic, and you can't add more event while delivering an event.)

In this case, the `sync` broadcast stream delivers an event *during* the `listen` call. It really shouldn't.
Sadly, there is existing code depending on that behavior, so it's likely going to be hard to change.

TL;DR: Don't add events before `onListen`. Don't add events during `onListen`.

**Comment by beeth0ven** [comment]
Thank you for clarification!

It helps me understand how it works.
1. A `sync` non-broadcast controller will buffer events before or during listen, and replay these events to listener after listened (using microtask)
2. A `sync` broadcast stream should deliver event "immediately" to listeners, if listener is empty then drop event 
3. It's not good to add events before or during listen

Ha,  as a  ReactiveX(Rx) fan I feel bellow feature is not going to be supported with dart stream, but I just note it here. Hoping this may be taken into consideration with stream in future 😄.
Observerable in Rx, this example is picked from rxjs docs: https://rxjs.dev/guide/observable

```js
import { Observable } from 'rxjs';
 
const observable = new Observable(subscriber => {
  subscriber.next(1);
  subscriber.next(2);
  subscriber.next(3);
  setTimeout(() => {
    subscriber.next(4);
    subscriber.complete();
  }, 1000);
});
 
console.log('just before subscribe');
observable.subscribe({
  next(x) { console.log('got value ' + x); },
  error(err) { console.error('something wrong occurred: ' + err); },
  complete() { console.log('done'); }
});
console.log('just after subscribe');
``` 

will prints:

```
just before subscribe
got value 1
got value 2
got value 3
just after subscribe
got value 4
done
```

Observable would deliver event during listen. I hope there is some way we can achieve it with dart stream.

Anyway! thanks for providing these useful tool.

Best wishes!

**Comment by lrhn** [comment]
Dart streams are not built for getting callbacks during `listen`.
You can call `listen` with not arguments (almost) and set the listeners on the returned `StreamSubscription` after `listen` returns. That is intended to be a *safe* way to initialize the listening. Any event sent during the listen` call will find no callback waiting for it, and the event is lost.

I know that Rx is used to doing something different, and Dart `Stream`s don't match well with that practice.

In practice, I'd recommend using a `Stream.multi` created stream for everything, and use the asynchronous event `add` method for events sent during `listen`, and `addSync` for events added after that (in response to other events, so it *looks* to the receiver like it's a new microtask, not in a way that can interrupt synchronous code).
Never use a stream to deliver an event "right now". Only use `sync` stream controllers to remove an *extra* asynchronous delay, not to remove the last one (just like you shouldn't have a `Future<Future<X>>`, you can always make that a plain `Future<X>`, but you can't remove the *last* `Future` from the type without actually waiting).

In practice, it probably works, if everybody holds the streams/listeners just right. It's just not something that's intended or guaranteed to be supported. Use with care.

**Comment by beeth0ven** [comment]
Thanks for taking the time to explain the situation we are in today.

I see `Stream.multi` is a recommended way to create stream. It make sense to me.

Since dart stream has it's own characteristic, with it's domain problems to be solved. I'm okay with this. Although I have words to futher discuss about the design, I'd like to keep them internal (if you really want to see, I'm not hesitate to share). Because I known I'm a rx guy, what I states may not be objective and fair. The stream system should not be identical to rx.

I would follow the best practice with dart stream. Above response dose a great job to help me understanding stream. Thanks again.

So feel free to close this issue since there is not much we can do with it right now,
Or keep it open for futher discuss and tracking.

Thank you.

**Comment by lrhn** [comment]
It is annoying that Stream and Rx are slightly incompatible. A stream is not just an iterable + asynchrony, and Dart made a choice early in the development of async (`Future`/`Stream`), that all events should be asynchonous events, so you, the user, had *any* chance of predicting when events could arrive. If we had had `await for` as the main way of accessing streams from the start, we might not have been so careful about allowing different ways to set up the listeners, but we didn't.

I feel your pain. I implemented streams, they are too damn complex, but every complexity was born out of an actual problem needing to be fixed. It would require an entirely different underlying model of asynchrony to change that.

**Comment by beeth0ven** [comment]
Thank you for your understanding!

Ha, you implemented streams. it's my honor to communicate with you. I agreed stream are too complex. I tried to implement it before. And  here is a naive implementation just for fun: https://dartpad.dev/?id=9dfd989bfe9ee41883661d47cac0c471&null_safety=true

Feel free to take a look 😄.

Have a good day!
